### ISSUE TITLE
PDO storage changes ATTR_ERRMODE

### ISSUE THREAD

**Issue by matias-larsson-primeo** [bshaffer/oauth2-server-php#813]
The module should not change the configuration of the PDO object (`$connection`) that is passed to it. This breaks the application's error handling when the same PDO instance is used elsewhere.

I'm talking about this line in `Pdo::__construct`
`$connection->setAttribute(\PDO::ATTR_ERRMODE, \PDO::ERRMODE_EXCEPTION);`

BR,
Matias Larsson

**Comment by afilippov1985** [comment]
Did you tried passing clone of `\PDO` object ?
like this
```php
$connection = new \PDO(....); // your connection
$storage = new OAuth2\Storage\Pdo(clone $connection);
$server = new OAuth2\Server($storage);
```

**Comment by matias-larsson-primeo** [comment]
Thanks for your reply.

I didn't try that originally because I would like to have just one mysql connection in the application. However, I was unable to find information about if the clone actually results in a new connection, so I tried. Turns out the clone causes my apache server to crash (SIGTERM). I haven't investigated further but it seems like it is not an option to clone PDO here.

BR,
Matias Larsson

**Comment by matias-larsson-primeo** [comment]
Hi,

One more correction: I think it's just the PHP process that segfaults, not the server.

I made a google search and found this: https://bugs.php.net/bug.php?id=49664
I think it could cause the segfault when cloning PDO, and is the reason why cloning PDO is out of question. Perhaps it is because PDO is implemented in the PHP sources and not as a PHP object.

Just wanted to share these points if anyone is looking into the issue.

**Comment by bluebaroncanada** [comment]
Do you have any code that you can provide?

**Comment by bluebaroncanada** [comment]
Agreed.  I can see other utility for this issue.  I'm actually going to mark this as an enhancement.  I want to leave the PDO class itself closed; however, I'm writing a new PDO MySql class and I will put this in it.  

You said it SEGFAULTs?

**Comment by matias-larsson-primeo** [comment]
Thanks.

Yes, when I clone the PDO instance, the PHP process seems to segfault. The script aborts and this is appended to apache error log:

```
[notice] child pid 56633 exit signal Segmentation fault (11)
```

Interestingly, it doesn't happen on the example I provided (by changing `new OAuth2\Storage\Pdo($pdo)` to `new OAuth2\Storage\Pdo(clone $pdo)` ), but does happen in the application code. In the application the PDO instance is owned by more than one object (as a class member) - other than that I don't know why it only crashes in the application. Must be that or something else that is done with the instance. I might look into this later, and if so, I will return with more info.

**Comment by bluebaroncanada** [comment]
That would be much appreciated.
